Game crashes when previewing Botania Multiblock

### Your GTNH Discord Username

Semetix

### Your Pack Version

2.8.0 - beta 1

### Your Server

Private Server

### Java Version

Java 21

### Type of Server

Don't know

### Your Actions

All Actions are seen in the video

### Crash Report

After that you need to restart your entire game otherwise you cant preview any receipe in NEI anymore
https://streamable.com/7x8eyp

[14:03:32] [Client thread/ERROR] [NotEnoughItems]: Error while looking up usage recipe
[14:03:32] [Client thread/ERROR] [NotEnoughItems]: inputId: item
[14:03:32] [Client thread/ERROR] [NotEnoughItems]: ingredients: [0xtile.automatedAlfPortal@0]
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: java.util.concurrent.ExecutionException: java.lang.NullPointerException
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.getException(ForkJoinTask.java:564)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:573)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1007)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.RecipeHandlerQuery.getHandlersWithRecipes(RecipeHandlerQuery.java:107)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.RecipeHandlerQuery.getRecipeHandlersParallel(RecipeHandlerQuery.java:67)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.RecipeHandlerQuery.runWithProfiling(RecipeHandlerQuery.java:48)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.GuiUsageRecipe.getUsageHandlers(GuiUsageRecipe.java:60)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.GuiUsageRecipe.openRecipeGui(GuiUsageRecipe.java:30)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.ItemHistoryPanel.mouseUp(ItemHistoryPanel.java:205)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.LayoutManager.onMouseUp(LayoutManager.java:249)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.guihook.GuiContainerManager.mouseUp(GuiContainerManager.java:543)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.guihook.GuiContainerManager.overrideMouseUp(GuiContainerManager.java:536)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.gui.inventory.GuiContainer.func_146286_b(GuiContainer.java)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.gui.GuiScreen.func_146274_d(GuiScreen.java:306)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.gui.inventory.GuiContainer.func_146274_d(GuiContainer.java)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.gui.GuiScreen.redirect$bpc000$modularui2$modularui$injectMouseInputEvent(GuiScreen.java:2548)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.gui.GuiScreen.func_146269_k(GuiScreen.java:268)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.Minecraft.func_71407_l(Minecraft.java:1640)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.Minecraft.func_71411_J(Minecraft.java:973)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:7610)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.client.main.Main.main(SourceFile:148)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at System//net.minecraft.launchwrapper.Launch.rfb$realLaunch(Launch.java:250)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at System//net.minecraft.launchwrapper.Launch.launch(Launch.java:35)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at System//net.minecraft.launchwrapper.Launch.main(Launch.java:60)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at com.gtnewhorizons.retrofuturabootstrap.Main.main(Main.java:207)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at System//org.prismlauncher.launcher.impl.StandardLauncher.launch(StandardLauncher.java:105)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at System//org.prismlauncher.EntryPoint.listen(EntryPoint.java:129)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at System//org.prismlauncher.EntryPoint.main(EntryPoint.java:70)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: Caused by: java.lang.NullPointerException
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.getException(ForkJoinTask.java:557)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	... 31 more
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: Caused by: java.lang.NullPointerException
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.getException(ForkJoinTask.java:557)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:573)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:663)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:677)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateParallel(ReduceOps.java:927)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:264)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:702)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.RecipeHandlerQuery.lambda$getHandlersWithRecipes$6(RecipeHandlerQuery.java:107)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask$AdaptedInterruptibleCallable.compute(ForkJoinTask.java:1689)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask$InterruptibleTask.exec(ForkJoinTask.java:1641)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1489)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2071)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2033)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: Caused by: java.lang.NullPointerException: Cannot invoke "net.minecraft.item.Item.func_77653_i(net.minecraft.item.ItemStack)" because the return value of "net.minecraft.item.ItemStack.func_77973_b()" is null
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//net.minecraft.item.ItemStack.func_82833_r(ItemStack.java:427)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//blockrenderer6343.integration.nei.GuiMultiblockHandler.getMultiblockName(GuiMultiblockHandler.java:426)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//blockrenderer6343.integration.nei.MultiblockHandler.getRecipeName(MultiblockHandler.java:169)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at Launch//codechicken.nei.recipe.RecipeHandlerQuery.lambda$getHandlersWithRecipes$3(RecipeHandlerQuery.java:96)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1709)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:546)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:759)
[14:03:32] [Client thread/INFO] [STDERR]: [java.lang.Throwable$WrappedPrintStream:println:807]: 	... 5 more
[14:03:32] [Client thread/INFO] [TabbyChat]: [CHAT] There was error during loading recipe. Please check log for details.

### Final Checklist

- [x] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying the crash still exists will prompt us to investigate and reopen it once we confirm your report.
- [x] I can reproduce this crash consistently by following the exact steps I described above.
- [ ] I have asked other people and they confirm they also crash by following the exact steps I described above.

--- a/dependencies.gradle
+++ b/dependencies.gradle
@@ -1,18 +1,18 @@
 // Add your dependencies here
 
 dependencies {
-    api('com.github.GTNewHorizons:Botania:1.12.12-GTNH:dev')
-    api('com.github.GTNewHorizons:GT5-Unofficial:5.09.51.357:dev')
+    api('com.github.GTNewHorizons:Botania:1.12.16-GTNH:dev')
+    api('com.github.GTNewHorizons:GT5-Unofficial:5.09.51.393:dev')
     api('thaumcraft:Thaumcraft:1.7.10-4.2.3.5:dev')
-    api("com.github.GTNewHorizons:StructureLib:1.4.12:dev")
-    api("com.github.GTNewHorizons:Avaritia:1.67:dev")
+    api("com.github.GTNewHorizons:StructureLib:1.4.15:dev")
+    api("com.github.GTNewHorizons:Avaritia:1.68:dev")
 
-    compileOnly('com.github.GTNewHorizons:NotEnoughItems:2.7.59-GTNH:dev')
+    compileOnly('com.github.GTNewHorizons:NotEnoughItems:2.7.64-GTNH:dev')
     compileOnly("com.github.GTNewHorizons:Chisel:2.16.5-GTNH:dev") {transitive = false}
 
     runtimeOnly('com.github.GTNewHorizons:Baubles-Expanded:2.1.9-GTNH:dev')
 
-    devOnlyNonPublishable('com.github.GTNewHorizons:NotEnoughItems:2.7.59-GTNH:dev')
+    devOnlyNonPublishable('com.github.GTNewHorizons:NotEnoughItems:2.7.64-GTNH:dev')
 }
 
 repositories {

--- a/src/main/java/net/fuzzycraft/botanichorizons/addons/Multiblocks.java
+++ b/src/main/java/net/fuzzycraft/botanichorizons/addons/Multiblocks.java
@@ -102,10 +102,10 @@ public final class Multiblocks {
 
     public static void postInit() {
         if (Mods.StructureLib.isModLoaded()) {
-            HoloProjectorSupport.registerWithStructureLib(poolInfusion, TileAdvancedCraftingPool.class);
-            HoloProjectorSupport.registerWithStructureLib(poolAlchemy, TileAdvancedAlchemyPool.class);
-            HoloProjectorSupport.registerWithStructureLib(poolConjuration, TileAdvancedConjurationPool.class);
-            HoloProjectorSupport.registerWithStructureLib(alfPortal, TileAdvancedAlfPortal.class);
+            HoloProjectorSupport.registerWithStructureLib(poolInfusion, BHBlocks.autoPoolInfusion, TileAdvancedCraftingPool.class);
+            HoloProjectorSupport.registerWithStructureLib(poolAlchemy, BHBlocks.autoPoolAlchemy, TileAdvancedAlchemyPool.class);
+            HoloProjectorSupport.registerWithStructureLib(poolConjuration, BHBlocks.autoPoolConjuration, TileAdvancedConjurationPool.class);
+            HoloProjectorSupport.registerWithStructureLib(alfPortal, BHBlocks.autoPortal, TileAdvancedAlfPortal.class);
         }
     }
 }

--- a/src/main/java/net/fuzzycraft/botanichorizons/util/structurelib/HoloProjectorSupport.java
+++ b/src/main/java/net/fuzzycraft/botanichorizons/util/structurelib/HoloProjectorSupport.java
@@ -9,6 +9,7 @@ import net.fuzzycraft.botanichorizons.util.multiblock.MetaBlockCheck;
 import net.fuzzycraft.botanichorizons.util.multiblock.MultiblockCheck;
 import net.fuzzycraft.botanichorizons.util.multiblock.MultiblockHelper;
 import net.fuzzycraft.botanichorizons.util.multiblock.MultiblockStructure;
+import net.minecraft.block.Block;
 import net.minecraft.tileentity.TileEntity;
 
 import java.util.Collections;
@@ -24,7 +25,7 @@ public class HoloProjectorSupport {
 
     public static final String HOLO_DEFAULT_CHANNEL = "main";
 
-    public static <T extends TileEntity> void registerWithStructureLib(MultiblockHelper definition, Class<T> tileClass) {
+    public static <T extends TileEntity> void registerWithStructureLib(MultiblockHelper definition, Block controllerBlock, Class<T> controllerTileClass) {
         final Map<MultiblockCheck, Character> keys = new HashMap<>();
         int xMin = definition.blocks[0].dx, xMax = definition.blocks[0].dx;
         int yMin = definition.blocks[0].dy, yMax = definition.blocks[0].dy;
@@ -45,6 +46,7 @@ public class HoloProjectorSupport {
                 counter++;
             }
         }
+        final String controllerChar = ((Character)counter).toString();
 
         // prepare definition string since we don't seem to have access to internals
         // note: stuff is too symmetric ATM, x/z indexing has not been fully tested
@@ -67,6 +69,11 @@ public class HoloProjectorSupport {
             structureString[zPos][yPos] = replaced;
         }
 
+        // insert controller block back into structure
+        final String controllerRowOld = structureString[-zMin][-yMin];
+        final String controllerReplaced = controllerRowOld.substring(0, -xMin) + controllerChar + controllerRowOld.substring(-xMin + 1);
+        structureString[-zMin][-yMin] = controllerReplaced;
+
         // Run everything through the StructureLib builder
         StructureDefinition.Builder<T> builder = IStructureDefinition.builder();
         builder.addShape(HOLO_DEFAULT_CHANNEL, structureString);
@@ -75,9 +82,11 @@ public class HoloProjectorSupport {
             IStructureElement<T> element = genStructureElement(check);
             builder.addElement(key, element);
         }
+        builder.addElement(counter, ofBlockAnyMeta(controllerBlock));
+
         IStructureDefinition<T> structure = builder.build();
         HoloContainer<T> container = new HoloContainer<>(structure, -xMin, -yMin, -zMin);
-        IMultiblockInfoContainer.registerTileClass(tileClass, container);
+        IMultiblockInfoContainer.registerTileClass(controllerTileClass, container);
     }
 
     private static <T> IStructureElement<T> genStructureElement(MultiblockCheck check) {
@@ -88,4 +97,5 @@ public class HoloProjectorSupport {
             return ofBlock(parsed.referenceBlock, parsed.referenceMeta);
         } else return null;
     }
+
 }

